sw/device/tests/dif/dif_gpio_test.cc

```diff
@@ -0,0 +1,476 @@
+}
+
+// Base class for the rest of the tests in this file, provides a
+// `dif_gpio_t` instance.
+class DifGpioTestInitialized : public DifGpioTest {
+ protected:
+  uint32_t AllZerosExcept(uint32_t index) { return 1 << index; }
+
+  uint32_t AllOnesExcept(uint32_t index) { return ~AllZerosExcept(index); }
+
+  const dif_gpio_t gpio_ = {.base_addr = dev().region()};
+};
+
+// Reset tests
+class ResetTest : public DifGpioTestInitialized {};
+
+TEST_F(ResetTest, NullArgs) {
+  EXPECT_EQ(dif_gpio_reset(nullptr), kDifGpioResultInvalidArgument);
+}
+
+TEST_F(ResetTest, Reset) {
+  EXPECT_WRITE32(GPIO_INTR_ENABLE_REG_OFFSET, 0);
+  EXPECT_WRITE32(GPIO_DIRECT_OE_REG_OFFSET, 0);
+  EXPECT_WRITE32(GPIO_DIRECT_OUT_REG_OFFSET, 0);
+  EXPECT_WRITE32(GPIO_INTR_CTRL_EN_RISING_REG_OFFSET, 0);
+  EXPECT_WRITE32(GPIO_INTR_CTRL_EN_FALLING_REG_OFFSET, 0);
+  EXPECT_WRITE32(GPIO_INTR_CTRL_EN_LVLHIGH_REG_OFFSET, 0);
+  EXPECT_WRITE32(GPIO_INTR_CTRL_EN_LVLLOW_REG_OFFSET, 0);
+  EXPECT_WRITE32(GPIO_CTRL_EN_INPUT_FILTER_REG_OFFSET, 0);
+  EXPECT_WRITE32(GPIO_INTR_STATE_REG_OFFSET, 0xFFFFFFFFu);
+
+  EXPECT_EQ(dif_gpio_reset(&gpio_), kDifGpioResultOK);
+}
+
+// Read tests
+class ReadTest : public DifGpioTestInitialized {};
+
+TEST_F(ReadTest, NullArgs) {
+  uint32_t out_arg_uint32_t;
+  bool out_arg_bool;
+
+  EXPECT_EQ(dif_gpio_all_read(nullptr, &out_arg_uint32_t),
+            kDifGpioResultInvalidArgument);
+  EXPECT_EQ(dif_gpio_all_read(&gpio_, nullptr), kDifGpioResultInvalidArgument);
+  EXPECT_EQ(dif_gpio_all_read(nullptr, nullptr), kDifGpioResultInvalidArgument);
+
+  EXPECT_EQ(dif_gpio_pin_read(nullptr, 0, &out_arg_bool),
+            kDifGpioResultInvalidArgument);
+  EXPECT_EQ(dif_gpio_pin_read(&gpio_, 0, nullptr),
+            kDifGpioResultInvalidArgument);
+  EXPECT_EQ(dif_gpio_pin_read(nullptr, 0, nullptr),
+            kDifGpioResultInvalidArgument);
+}
+
+TEST_F(ReadTest, AllPins) {
+  const uint32_t exp_pin_values = 0xA5A5A5A5;
+  EXPECT_READ32(GPIO_DATA_IN_REG_OFFSET, exp_pin_values);
+
+  uint32_t pin_values = 0;
+  EXPECT_EQ(dif_gpio_all_read(&gpio_, &pin_values), kDifGpioResultOK);
+  EXPECT_EQ(pin_values, exp_pin_values);
+}
+
+TEST_F(ReadTest, SinglePin) {
+  for (uint32_t pin = 0; pin < 32; ++pin) {
+    for (const bool exp_pin_val : {true, false}) {
+      const uint32_t reg_val =
+          exp_pin_val ? AllZerosExcept(pin) : AllOnesExcept(pin);
+      EXPECT_READ32(GPIO_DATA_IN_REG_OFFSET, reg_val);
+      bool pin_val = !exp_pin_val;
+      EXPECT_EQ(dif_gpio_pin_read(&gpio_, pin, &pin_val), kDifGpioResultOK);
+      EXPECT_EQ(pin_val, exp_pin_val);
+    }
+  }
+}
+
+// Write tests
+class WriteTest : public DifGpioTestInitialized {};
+
+TEST_F(WriteTest, NullArgs) {
+  EXPECT_EQ(dif_gpio_all_write(nullptr, 0xFFFFFFFF),
+            kDifGpioResultInvalidArgument);
+  EXPECT_EQ(dif_gpio_pin_write(nullptr, 0, true),
+            kDifGpioResultInvalidArgument);
+  EXPECT_EQ(dif_gpio_masked_write(nullptr, 0xFFFFFFFF, 0xFFFFFFFF),
```

Comment:
In previous reviews @mcy prefered to use `std::numeric_limits<T>::max` instead of `0xFFFFFFFF`. I guess could be done something like `constexpr kAllBitsSet = std::numeric_limits<uint32_t>::max;`, if you would like to avoid long lines.

Reply:
Done.